### ISSUE TITLE
Clarify documentation around refresh_session_interval vs access_token_expires_in

### ISSUE THREAD

**Issue by edmorley** [pingidentity/lua-resty-openidc#66]
The readme currently says:
```
    -- Refresh the user's id_token after 900 seconds without requiring re-authentication
    --refresh_session_interval = 900,
    ...
    --access_token_expires_in = 3600
    -- Default lifetime in seconds of the access_token if no expires_in attribute is present in the token 
    endpoint response.
       This plugin will silently renew the access_token once it's expired if refreshToken scope is present.
```

I interpreted this to mean:
* If less than 900 seconds have passed, don't validate anything, reuse the existing session.
* If between 900 and 3600 seconds have passed, silently refresh the id_token/access token server side using the refresh token, but with no client side redirects.
* If >3600 seconds have passed, then redirect the client to the OP for re-authentication.

Whereas from reading the source the behaviour is actually:
* If less than 900 seconds have passed, don't validate anything, reuse the existing session.
* If >900 force refresh of the access token by redirecting the client to the OP with `prompt = "none"`.
(Presuming the access token returned from the OP doesn't have an expiry of its own which is shorter than 3600 seconds)

If I've interpreted the behaviour correctly, it would seem the `refresh_session_interval` description in the docs might be more accurately worded as:
> Redirect the client to the OP (but with no authentication prompt) after 900 seconds to refresh their id_token and access token.

Also the `access_token_expires_in` description should probably read:
> ...
> This plugin will silently renew the access_token once it's expired if refreshToken scope is present, **so long as `refresh_session_interval` is not set**.

Many thanks!

**Comment by zandbelt** [comment]
thanks for the comments, I'd like to see @gonzalad comment on it

**Comment by edmorley** [comment]
Actually thinking about this, I wonder whether the behaviour of redirecting to the OP with prompt=none is correct at all? It seems part of the purpose of that is to revalidate the user, but can't the access token by used server side to do that instead?

**Comment by zandbelt** [comment]
one may want to verify that the user still has an ongoing SSO session with the Provider; that's what `prompt=none` is for

**Comment by gonzalad** [comment]
Hello,

Please bear in mind that I'm know more the refresh_session_interval parameter.

**About refresh_session_interval**

This setting controls access_token, id_token and refresh_token renewal by redirecting the end-user to the OP (which will silently reauthenticate the end-user).

> If I've interpreted the behaviour correctly, it would seem the `refresh_session_interval` description in the docs might be more accurately worded as: 
> Redirect the client to the OP (but with no authentication prompt) after 900 seconds to refresh their id_token and access token.

Exactly. This redirect is done whatever the value of access_token_expires_in.
And the RP gets new id_token, access token (and eventually refresh_token).

**About access_token_expires_in**

This setting silently refreshes the access_token whenever it's expired without any end-user intervention (no end-user redirect).

> This plugin will silently renew the access_token once it's expired if refreshToken scope is present, so 
> long as refresh_session_interval is not set.

If refresh_session_interval is used and its value is less than accessToken expires_in,  accessToken refresh will never be triggered.
If refresh_session_interval value is greater than accessToken expires_in 
I'd say `This plugin will silently renew the access_token once it's expired if refreshToken scope is present.`


**More explanation**

To try to better explain, I'll show you my current use case of access_token_expires_in.
In my setup:
 - my Authorization Server issues tokens with a very short lifetime (3 minutes) : its token endpoint returns an accessToken with expires_in=180.
 - I don't use refresh_session_interval functionality.

Here's my settings
```
local opts = {
    redirect_uri_path = "/redirect_uri",
    discovery = "http://authorization-server/oidc/.well-known/openid-configuration",
    client_id = "xxx",
    client_secret = "yyy",
    scope = "openid refreshToken",
    iat_slack = 600,
    redirect_uri_scheme = "http",
    logout_path = "/logout",
    token_endpoint_auth_method = "client_secret_basic",
    ssl_verify = "no"
}
```

- The end-user goes to RP (nginx lua)
 - RP redirects to OP
 - end user authenticates in OP and is redirected to RP
 - RP retrieves accessToken, idToken, refreshToken by calling OP token endpoint.
 - End-user browser calls some REST Services (ajax) through the RP
- The RP propagates the accessToken to the REST Service (RS).

_What would happen without access_token_expires_in feature ?_
3 minutes after being authenticated, the nginx would have called the REST Service with expired accessToken, and the REST Service would have returned a 401.

_What happens with access_token_expires_in feature ?_
3 minutes after being authenticated (minus a little leeway), nginx calls the token endpoint to get a new valid accessToken (the end-user isn't involved in this call).
The REST Service is called with a valid accessToken.

_What happens when nginx session is expired ?_
i.e. if end-user doesn't use the nginx oidc app for more than 1 hour, then nginx session is expired.
Next time the end-user access the RP, the RP redirects him to OP.
 - If end-user is authenticated to OP (if his OP session isn't expired), then he's silently reauthenticated (at least in my OP setup where I don't show the grant scope screenà.
 - If end-user isn't anymore authenticated to the OP (OP session expired), then he must submit his credentials to authenticate.
 
_Why would I use refresh_session_interval ?_
Once more, I don't use this setting, so perhaps (and surely) I'm missing something here.
The only case I see would be to enforce that the Authorization Server session isn't destroyed.

**Comment by edmorley** [comment]
Yeah this matches what I thought -- that `refresh_session_interval` is slightly redundant.

In addition to the changes above, it may also be worth mentioning lua-resty-session's `session_cookie_lifetime`, which defaults to 3600 (so if any of the expiry times set here are longer, the session will be lost anyway).

**Comment by gdestuynder** [comment]
Note: the key difference between `access_token_expires_in` and `refresh_session_interval` is that `refresh_session_interval` does **not** require a refresh token.
Refresh tokens are inherently dangerous since they are tokens which allow creating access forever.

Depending on your use case or requirements one or the other might be better.

That being said I agree we could make it more clear in the docs :)
